File: output/GPT3.5-Turbo/xCodeEval_remove_4000/Java/Go/dff2744b9701a10438fa6c925781d1b5.go
package main

import (
	"bufio"
	"fmt"
	"io"
	"os"
	"sort"
)

type Graph struct {
	vertexCount int
	edgeCount   int
	firstOut    []int
	from        []int
	to          []int
	nextOut     []int
	flags       []int
}

func NewGraph(vertexCount, edgeCapacity int) *Graph {
	g := &Graph{
		vertexCount: vertexCount,
		edgeCount:   0,
		firstOut:    make([]int, vertexCount),
		from:        make([]int, edgeCapacity),
		to:          make([]int, edgeCapacity),
		nextOut:     make([]int, edgeCapacity),
		flags:       make([]int, edgeCapacity),
	}
	for i := 0; i < vertexCount; i++ {
		g.firstOut[i] = -1
	}
	return g
}

func (g *Graph) AddEdge(from, to int) {
	g.from[g.edgeCount] = from
	g.to[g.edgeCount] = to
	g.nextOut[g.edgeCount] = g.firstOut[from]
	g.firstOut[from] = g.edgeCount
	g.edgeCount++
}

type TaskD struct {
	graph    *Graph
	weight   []int
	path     []int
}

func (task *TaskD) Solve(input *InputReader, output *OutputWriter) {
	n := input.ReadInt()
	k := input.ReadInt()
	a := make([]int, n)
	for i := 0; i < n; i++ {
		a[i] = input.ReadInt()
	}
	u := make([]int, n-1)
	v := make([]int, n-1)
	for i := 0; i < n-1; i++ {
		u[i] = input.ReadInt() - 1
		v[i] = input.ReadInt() - 1
	}
	graph := NewGraph(n, 2*(n-1))
	for i := 0; i < n-1; i++ {
		graph.AddEdge(u[i], v[i])
		graph.AddEdge(v[i], u[i])
	}
	order := make([]int, n)
	for i := 0; i < n; i++ {
		order[i] = i
	}
	sort.Slice(order, func(i, j int) bool {
		return a[order[i]] > a[order[j]]
	})
	left := k
	right := n
	weight := make([]int, n)
	degree := make([]int, n)
	for i := 0; i < n-1; i++ {
		degree[u[i]]++
		degree[v[i]]++
	}
	remaining := make([]int, n)
	queue := make([]int, n)
	path := make([]int, n)
	for left < right {
		middle := (left + right) >> 1
		threshold := a[order[middle-1]]
		for i := 0; i < n; i++ {
			if a[i] >= threshold {
				weight[i] = 1
			} else {
				weight[i] = 0
			}
		}
		copy(remaining, degree)
		size := 0
		for i := 0; i < n; i++ {
			if degree[i] == 1 && weight[i] > 0 {
				queue[size] = i
				size++
			}
		}
		for i := 0; i < size; i++ {
			current := queue[i]
			for j := graph.firstOut[current]; j != -1; j = graph.nextOut[j] {
				next := graph.to[j]
				if weight[next] != 0 {
					weight[next] += weight[current]
					weight[current] = 0
					remaining[next]--
					if remaining[next] == 1 {
						queue[size] = next
						size++
					}
				}
			}
		}
		result := task.dfs(0, -1, graph, weight, path)
		if result >= k {
			right = middle
		} else {
			left = middle + 1
		}
	}
	output.PrintLine(a[order[left-1]])
}

func (task *TaskD) dfs(current, last int, graph *Graph, weight, path []int) int {
	result := 0
	max := 0
	second := 0
	for i := graph.firstOut[current]; i != -1; i = graph.nextOut[i] {
		next := graph.to[i]
		if next == last {
			continue
		}
		result = maxInt(result, task.dfs(next, current, graph, weight, path))
		if path[next] > max {
			second = max
			max = path[next]
		} else {
			second = maxInt(second, path[next])
		}
	}
	if weight[current] != 0 {
		path[current] = weight[current] + max
		result = maxInt(result, max+second+weight[current])
	} else {
		path[current] = 0
	}
	return result
}

func maxInt(a, b int) int {
	if a > b {
		return a
	}
	return b
}

type InputReader struct {
	reader *bufio.Reader
}

func NewInputReader(r io.Reader) *InputReader {
	return &InputReader{reader: bufio.NewReader(r)}
}

func (in *InputReader) ReadInt() int {
	var x int
	_, _ = fmt.Fscan(in.reader, &x)
	return x
}

type OutputWriter struct {
	writer *bufio.Writer
}

func NewOutputWriter(w io.Writer) *OutputWriter {
	return &OutputWriter{writer: bufio.NewWriter(w)}
}

func (out *OutputWriter) PrintLine(a ...interface{}) {
	_, _ = fmt.Fprintln(out.writer, a...)
}

func (out *OutputWriter) Flush() {
	_ = out.writer.Flush()
}

func main() {
	input := NewInputReader(os.Stdin)
	output := NewOutputWriter(os.Stdout)
	task := &TaskD{}
	task.Solve(input, output)
	output.Flush()
}

File: output/GPT4o-Mini/xCodeEval_4000/Java/C/dff2744b9701a10438fa6c925781d1b5.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <limits.h>

typedef struct {
    int *data;
    int size;
} IntArray;

typedef struct {
    int *firstOutbound;
    int *nextOutbound;
    int *from;
    int *to;
    int edgeCount;
    int vertexCount;
} Graph;

typedef struct {
    int *weight;
    int *path;
    Graph *graph;
} TaskD;

void readIntArray(IntArray *arr, int size) {
    arr->data = (int *)malloc(size * sizeof(int));
    arr->size = size;
    for (int i = 0; i < size; i++) {
        scanf("%d", &arr->data[i]);
    }
}

void createGraph(Graph *graph, int vertexCount, int edgeCapacity) {
    graph->vertexCount = vertexCount;
    graph->edgeCount = 0;
    graph->firstOutbound = (int *)malloc(vertexCount * sizeof(int));
    graph->nextOutbound = (int *)malloc(edgeCapacity * sizeof(int));
    graph->from = (int *)malloc(edgeCapacity * sizeof(int));
    graph->to = (int *)malloc(edgeCapacity * sizeof(int));
    memset(graph->firstOutbound, -1, vertexCount * sizeof(int));
}

void addEdge(Graph *graph, int fromID, int toID) {
    graph->from[graph->edgeCount] = fromID;
    graph->to[graph->edgeCount] = toID;
    graph->nextOutbound[graph->edgeCount] = graph->firstOutbound[fromID];
    graph->firstOutbound[fromID] = graph->edgeCount++;
}

int dfs(TaskD *task, int current, int last) {
    int result = 0;
    int max = 0;
    int second = 0;
    for (int i = task->graph->firstOutbound[current]; i != -1; i = task->graph->nextOutbound[i]) {
        int next = task->graph->to[i];
        if (next == last) continue;
        result = dfs(task, next, current);
        if (task->path[next] > max) {
            second = max;
            max = task->path[next];
        } else {
            second = (task->path[next] > second) ? task->path[next] : second;
        }
    }
    if (task->weight[current] != 0) {
        task->path[current] = task->weight[current] + max;
        result = (result > max + second + task->weight[current]) ? result : (max + second + task->weight[current]);
    } else {
        task->path[current] = 0;
    }
    return result;
}

void solve(TaskD *task, int n, int k, IntArray *a, IntArray *u, IntArray *v) {
    task->graph = (Graph *)malloc(sizeof(Graph));
    createGraph(task->graph, n, n - 1);
    for (int i = 0; i < n - 1; i++) {
        addEdge(task->graph, u->data[i], v->data[i]);
    }
    task->weight = (int *)calloc(n, sizeof(int));
    task->path = (int *)calloc(n, sizeof(int));
    int left = k, right = n;

    while (left < right) {
        int middle = (left + right) >> 1;
        int threshold = a->data[middle - 1];
        memset(task->weight, 0, n * sizeof(int));
        for (int i = 0; i < n; i++) {
            task->weight[i] = (a->data[i] >= threshold) ? 1 : 0;
        }
        int result = dfs(task, 0, -1);
        if (result >= k) {
            right = middle;
        } else {
            left = middle + 1;
        }
    }
    printf("%d\n", a->data[left - 1]);
}

int main() {
    int n, k;
    scanf("%d %d", &n, &k);
    IntArray a, u, v;
    readIntArray(&a, n);
    u.size = n - 1;
    u.data = (int *)malloc(u.size * sizeof(int));
    v.size = n - 1;
    v.data = (int *)malloc(v.size * sizeof(int));
    readIntArray(&u, u.size);
    readIntArray(&v, v.size);

    TaskD task;
    solve(&task, n, k, &a, &u, &v);

    free(a.data);
    free(u.data);
    free(v.data);
    free(task.weight);
    free(task.path);
    free(task.graph->firstOutbound);
    free(task.graph->nextOutbound);
    free(task.graph->from);
    free(task.graph->to);
    free(task.graph);
    return 0;
}
